docs/i18n/20-i18n-dojo-apps.md

```diff
@@ -0,0 +1,213 @@
+
+### `I18nMixin` Widget Properties
+
+- `locale`?: string
+    -  The locale for the widget.<br>If not specified, then the [root application locale](#providing-locale-data-to-i18n-aware-widgets) or [its override](#changing-locales) is assumed.<br>If specified, the widget's DOM node will have a `lang` property set to the locale.
+- `rtl`?: boolean
+    - An optional flag indicating the widget's text direction. If `true`, then the underlying DOM node's `dir` property is set to `"rtl"`. If it is `false`, then the `dir` property is set to `"ltr"`. Otherwise, the property is not set.
+- `i18nBundle`?: `Bundle<Messages>` | `Map<Bundle<Messages>, Bundle<Messages>>`
+    -  An optional override for the [default language bundle](./10-bundles.md#default-language-module) passed to the `localizeBundle` method. If the override contains a `messages` object, then it will completely replace the underlying default language bundle that the widget may be using. If the override only contains a `locales` object, a new bundle will be created with the additional locale loaders specified in the override.
+
+### `I18nMixin` `localizeBundle()` method
+
+Widgets can pass in their [default language bundle](./10-bundles.md#default-language-module) into the `localizeBundle` method to have the bundle localized appropriately given the widget's `locale` property.
+
+If the bundle supports the widget's current locale, but those locale-specific messages have not yet been loaded, then a bundle of blank message values is returned. Alternatively, the `localizeBundle` method accepts a second boolean argument, which, when `true`, causes the default messages to be returned instead of the blank bundle. The widget will be invalidated once the locale-specific messages have been loaded, triggering a re-render with the localized message content.
+
+The object returned by `localizeBundle` contains the following properties and methods:
+
+-   `messages`: An object containing the localized message key-value pairs. If the messages have not yet loaded, then `messages` will be either a blank bundle or the default messages, depending upon how `localizeBundle` was called.
+-   `isPlaceholder`: a boolean property indicating whether the returned messages are the actual locale-specific messages (`false`) or just the placeholders used while waiting for the localized messages to finish loading (`true`). This is useful to prevent the widget from rendering at all if localized messages have not yet loaded.
+-   `format(key: string, replacements: { [key: string]: string })`: a method that accepts a message key as its first argument and an object of replacement values as its second. For example, if the bundle contains `greeting: 'Hello, {name}!'`, then calling `format('greeting', { name: 'World' })` would return `'Hello, World!'`.
+
+An example of using all features returned by `localizeBundle`:
+```ts
+const MyWidgetBase = I18nMixin(WidgetBase);
```

Comment:
We can just use `I18nMixin` in line now, i.e.

```ts
class I18nWidget extends I18nMixin(WidgetBase)<I18nWidgetProperties> {
```

Reply:
Also do we want all the examples to include the imports/interfaces etc required to be able to copy and paste a working example?

Reply:
Fixed & tested as working